generic_config_updater/patch_sorter.py

```diff
@@ -333,6 +333,114 @@ def _extend_moves(self, move, diff):
 
+class ConfigFilter:
+    """
+    A filtering class to get the paths matching the filter from the given config.
+    The patterns:
+    - Each pattern consist of multiple tokens
+    - Tokens are matched with the config from the root level
+    - Each token can be:
+      - '*' Will match all keys at the current level
+      - '@' Will be replaced by 'common_key' passed in 'get_paths'
+      - <other> Will match other strings that are not * nor @
+    - Token advanced options
+      - '*|@' Will match keys that end with '|common_key'
+      - '@|*' Will match keys that start with 'common_key|'
+      - '*|<other>' Will match keys that end with '|<other>'
+      - '<other>|*' Will match keys that start with '<other>|'
+    """
+    def __init__(self, patterns, path_addressing):
+        self.patterns = patterns
+        self.path_addressing = path_addressing
+
+    def get_paths(self, config, common_key=None):
+        for pattern in self.patterns:
+            for path in self._get_paths_recursive(config, pattern, [], 0, common_key):
+                yield path
+
+    def _get_paths_recursive(self, config, pattern_tokens, matching_tokens, idx, common_key):
+        if idx == len(pattern_tokens):
+            yield self.path_addressing.create_path(matching_tokens)
+            return
+
+        token = pattern_tokens[idx]
+        if common_key:
+            token = token.replace("@", common_key)
+
+        matching_keys = []
+        if token == "*":
+            matching_keys = config.keys()
+        elif token.startswith("*|"):
+            suffix = token[2:]
+            matching_keys = [key for key in config.keys() if key.endswith(suffix)]
+        elif token.endswith("|*"):
+            prefix = token[:-2]
+            matching_keys = [key for key in config.keys() if key.startswith(prefix)]
+        elif token in config:
+            matching_keys = [token]
+
+        for key in matching_keys:
+            matching_tokens.append(key)
+            for path in self._get_paths_recursive(config[key], pattern_tokens, matching_tokens, idx+1, common_key):
+                yield path
+            matching_tokens.pop()
+
+class PortCriticalConfigIdentifier:
+    def __init__(self, path_addressing):
+        # TODO: port-critical fields are hard-coded for now, it should be moved to YANG models
+        # NOTE: PortCritical validation logic can be generalized to be validation for any config that
+        #       requires another config to be of a specific value. Call the validator "TransitionalValueValidator"
+        #       For the case of port-critical validation the transitional value would be "admin_status=down".
+        self.port_critical_filter = ConfigFilter([
+                ["BUFFER_PG", "@|*"],
+                ["BUFFER_QUEUE", "@|*"],
+                ["QUEUE", "@|*"]
+            ],
+            path_addressing)
+        self.path_addressing = path_addressing
+
+    def identify_admin_up_ports(self, config):
```

Comment:
>identify_admin_up_ports [](http://example.com/codeflow?start=8&length=23)

Could you try leverage JsonPath, which is a mature query language? For this function, you may query `$.PORT.*[?(@.admin_status == 'up')]`.

ref: [https://support.smartbear.com/alertsite/docs/monitors/api/endpoint/jsonpath.html](https://support.smartbear.com/alertsite/docs/monitors/api/endpoint/jsonpath.html)
online evaluator: [http://jsonpath.herokuapp.com/?path=$..book[?(@.author%20=~%20/.*REES/i](http://jsonpath.herokuapp.com/?path=$..book[?(@.author%20=~%20/.*REES/i))]
python lib: [https://github.com/json-path/JsonPath#path-examples](https://github.com/json-path/JsonPath#path-examples)

Reply:
This is a JAVA lib, I think you meant: https://github.com/kennknowles/python-jsonpath-rw

Reply:
Check this reply[https://github.com/Azure/sonic-utilities/pull/1998/files/e957f7e072ccfa1cabc58eefd9219a1c53910098#r783290272]

Specifically for this case, I can add `get_values()` to `ConfigFilter` class.

In a future PR, I can leverage JsonPath instead the `Patterns`

Reply:
Future PR is good to me. Sorry for wrong library. I did not search for all the python libraries. python-jsonpath-rw seems a good candidate.